ONSdigital/rm-case-service
# Fix warnings on valid state transitions

# What and why?

There are a number of valid events that happen in the system that cause case to try and transition to a state that it thinks is invalid. This PR will try to fix a number of them as they're completely allowed within the system, but they've not been added to the state transition machine for one reason or another.

The 4 transitions that would give false warnings are:
- event=EQ_LAUNCH; from=INPROGRESS
- event=RESPONDENT_ENROLED; from=NOTSTARTED
- event=ACCESS_CODE_AUTHENTICATION_ATTEMPT from=NOTSTARTED
- event=ACCESS_CODE_AUTHENTICATION_ATTEMPT from=INPROGRESS

# How to test?

Run the acceptance tests to seed some data, this'll make your life a lot easier!

event=EQ_LAUNCH; from=INPROGRESS; to=INPROGRESS
- Launch an EQ (state should go from NOTSTARTED to INPROGRESS)
- Go back to frontstage then launch the EQ again.  Shouldn't give a warning anymore and should remain in the INPROGRESS state.

event=RESPONDENT_ENROLED; from=NOTSTARTED; to=NOTSTARTED
- As a respondent, enrol on a survey you're not enrolled on yet (the seeded data should give you 7 others to choose from, just grab an enrolment code from another reporting unit).  This used to cause an error but now should just happily stay in the NOTSTARTED state with no warnings

event=ACCESS_CODE_AUTHENTICATION_ATTEMPT from=NOTSTARTED; to=NOTSTARTED
- Go to the reporting unit 49900000001.  Click on the QBS survey.  Before case would've thrown a warning here but now it should stay in the NOTSTARTED state with no warnings

event=ACCESS_CODE_AUTHENTICATION_ATTEMPT from=INPROGRESS; to=INPROGRESS
- Sign in as the example user and start the QBS survey (the case should go to INPROGRESS)
- Go to the reporting unit 49900000001.  Click on the QBS survey.  Before case would've thrown a warning here but now it should stay in the NOTSTARTED state with no warnings

# Trello
